### PR TITLE
py/objarray: Add byte operations to bytearray.

### PR DESCRIPTION
**Pull request by andrewleech**
Reuse the implementations for bytes operations with updates to match cpython.
For details see https://docs.python.org/3/library/stdtypes.html#bytes-and-bytearray-operations

This started from a need for `find` function on a large bytearray without needing to copy it into a `byte()` first.  Others have needed this too https://forum.micropython.org/viewtopic.php?t=7967

Considering cpython supports these functions on both bytes and bytearray it makes sense to mirror them here.

### CODE REVIEW

**Comment by glenn20** [comment]
Very very happy to see this. Thanks.

File: py/objstr.c [hunk 1]
```diff
@@ -469,8 +487,15 @@ STATIC mp_obj_t str_join(mp_obj_t self_in, mp_obj_t arg) {
...
+    if (self_type == &mp_type_bytearray) {
+        self_type = &mp_type_bytes;
+    }
     for (size_t i = 0; i < seq_len; i++) {
-        if (mp_obj_get_type(seq_items[i]) != self_type) {
+        const mp_obj_type_t *seq_type = mp_obj_get_type(seq_items[i]);
+        if (seq_type == &mp_type_bytearray) {
+            seq_type = &mp_type_bytes;
+        }
+        if (seq_type != self_type) {
```
Comment (musteresel) [inline py/objstr.c:498]: I think you're missing a feature gate here? Causing the minimal build to fail.
~~~
#ifdef MICROPY_PY_BUILTINS_BYTEARRAY
~~~

Comment (andrewleech) [reply py/objstr.c:498]: Oh I thought I'd fixed that already... might have force pushed away that fix. I'll certainly need to fix that

File: py/objstr.c [hunk 2]
```diff
@@ -41,6 +41,23 @@ STATIC mp_obj_t str_modulo_format(mp_obj_t pattern, size_t n_args, const mp_obj_
 STATIC mp_obj_t mp_obj_new_bytes_iterator(mp_obj_t str, mp_obj_iter_buf_t *iter_buf);
 STATIC NORETURN void bad_implicit_conversion(mp_obj_t self_in);
 
+STATIC void mp_check_compat_arg(const mp_obj_type_t *self_type, const mp_obj_t arg) {
```
Comment (jimmo) [inline py/objstr.c:44]: As it's a static helper, shouldn't be called `mp_`. Maybe just `str_check_arg_type` or something.

File: tests/basics/bytes_count.py [hunk 3]
```diff
@@ -52,3 +52,5 @@ def t():
     return True
 
 print(b"0000".count(b'0', t()))
+
+print(bytearray(b"aaaa").count(b"a"))
```
Comment (jimmo) [inline tests/basics/bytes_count.py:56]: Newlines at the end of all four files (too many on bytearray_byte_operations.py, missing on the other three).

File: py/objstr.c (hunk 2 above)
Comment (jimmo) [reply py/objstr.c:44]: Maybe a quick comment just to explain what's going on here:

`// Arguments to string operations need to be the same type (e.g. str.find(str), with the exception that bytes may be used for bytearray and vice versa."

File: py/objstr.c [hunk 4]
```diff
@@ -2080,6 +2106,11 @@ mp_obj_t mp_obj_new_str_from_vstr(const mp_obj_type_t *type, vstr_t *vstr) {
     ((byte *)o->data)[o->len] = '\0'; // add null byte
     vstr->buf = NULL;
     vstr->alloc = 0;
+    #if MICROPY_PY_BUILTINS_BYTEARRAY
+    if (ret_type == &mp_type_bytearray) {
+        return mp_obj_new_bytearray_by_ref(o->len, (byte *)o->data);
```
Comment (jimmo) [inline py/objstr.c:2111]: It doesn't really make sense to construct the str/bytes and then abandon it just to use the data.

```
const byte *data;
if (vstr->len + 1 == vstr->alloc) {
    data = vstr->buf;
} else {
    data = m_renew(char, vstr->buf, vstr->alloc, vstr->len + 1);
}
data[len] = '\0'; // add null byte
vstr->buf = NULL;
vstr->alloc = 0;

#if MICROPY_PY_BUILTINS_BYTEARRAY
if (ret_type == &mp_type_bytearray) {
    return mp_obj_new_bytearray_by_ref(vstr->len, data);
}
#endif
mp_obj_str_t *o = mp_obj_malloc(mp_obj_str_t, type);
o->len = vstr->len;
o->hash = qstr_compute_hash(data, vstr->len);
o->data = data;
return MP_OBJ_FROM_PTR(o);
```

**Comment by jimmo** [comment]
This is +352 bytes on PYBV11, of which 192 bytes is the additional locals added to bytearray.

I think this is a good feature and worth adding.

One quirk worth noting is that this will also add all the methods to `array.array`, but (a) they aren't in CPython, and (b) they don't work (i.e. `array('i', [1,2,3]).find(array('i', [2]))` --> -1).

**Comment by jimmo** [comment]
The thing about array.array made me realise that you can potentially avoid duplicating the locals dict by making bytearray implement attr, and then forward to str8_locals_dict. It can't use `str8_locals_dict` directly as its locals because it also needs append/extend. This also means we can remove the (incorrect) decode method from array.

By doing this I can get this PR down to +208 bytes.

While we're here though, on a unicode build (most boards!), str (struni_locals_dict) and bytes (str8_locals_dict) are completely identical except that one has encode and the other has decode. If we make str and bytes/bytearray both have encode/decode then we can save -224 bytes by using the same table for str/bytes/bytearray.

We can do the same trick with attr, and use attr to add encode to str and decode to bytes/bytearray. This gives a saving of -120 bytes instead (which would put this PR at +88 bytes).

It's possibly a consideration to add .attr to str/bytes/bytearray though as these are frequently used classes, but all the methods we're talking about are already not cheap, and the .attr mechanism is pretty efficient.

**Comment by jimmo** [comment]
Here's a commit with the above ideas implemented: https://github.com/micropython/micropython/commit/07bd840964261adfc36297270606931008214cba

**Comment by jimmo** [comment]
I realised this morning that array/bytearray/bytes/str can actually share the exact same locals dict, as long as you order it in the right way and can offset the start and length.

Here's an implementation of that. https://github.com/micropython/micropython/commit/7a5136e81e91644e565e37e0d42c75a583a738ae It brings this PR overall to -40 bytes.

**Comment by andrewleech** [comment]
Thanks @jimmo that's awesome, love the idea. I've integrated that here as discussed and fixed the rest of the comments.

**Comment by dpgeorge** [comment]
This is a good improvement, but it only works by accident because the layout of `mp_obj_str_t` and `mp_obj_array_t` are very similar, they have the `len` and `data`/`items` entries in the same locations.  And so `GET_STR_DATA_LEN(...)` just so happens to work on bytearray objects.

But... it doesn't work for nanboxing because these structs are subtly different there.  That can be fixed by changing `mp_uint_t hash` to `size_t hash`, in `mp_obj_str_t`.

It's very fragile though to rely on these structs being the same!  We need to somehow enforce that they are the same, but it's not clear how to do it.

File: tests/basics/string_center.py [hunk 5]
```diff
@@ -11,3 +11,5 @@
 print("foo".center(5))
 print("foo".center(6))
 print("foo".center(20))
+
+print(b"foo".center(20))
```
Comment (dpgeorge) [inline tests/basics/string_center.py:15]: please put this test in a separate file `bytearray_center.py`, so it can be automatically skipped if the port doesn't have `bytearray`

Comment (andrewleech) [reply tests/basics/string_center.py:15]: I've just done that in the latest push, it's just a byte version I added left here how.  Is that better not having in a str_ test file?

Comment (dpgeorge) [reply tests/basics/string_center.py:15]: oh, sorry, I didn't see the update!

You'll see that there are specific tests for bytes methods, eg `bytes_strip.py`.  So it would be more consistent to have a new `bytes_center.py` test as well.

**Comment by andrewleech** [comment]
> It's very fragile though to rely on these structs being the same! We need to somehow enforce that they are the same, but it's not clear how to do it.

If I've understood the issue correctly, it's crucial that the len element in both structs has the same offset within the struct? If so this test might cover it? https://github.com/micropython/micropython/pull/9002/commits/f20946b1829b83c59dd268f34a848251fb475864

**Comment by dpgeorge** [comment]
That does test the correct thing.  But it might be more robust to add a new helper, like `MP_GET_STR_BYTES_BYTEARRAY_LEN_DATA` which explicitly handles str/bytes/bytearray, and then hope the compiler will optimise it to be the same as the existing `GET_STR_DATA_LEN`.

**Comment by andrewleech** [comment]
Ah yes I see the root problem now. Wouldn't a replacement MP_GET_STR_BYTES_BYTEARRAY_LEN_DATA stye function require a type check like?:
```
if (mp_obj_get_type(arg) == &mp_type_bytearray || mp_obj_get_type(arg) == &mp_type_array) {
    str_len = (mp_obj_array_t *)MP_OBJ_TO_PTR(str_obj_in))->len
} else {
str_len = ((mp_obj_str_t *)MP_OBJ_TO_PTR(str_obj_in))->len
```
In this case it'd never really be able to optimise down surely.

**Comment by dpgeorge** [comment]
Yes that's what it would need to do.  If `mp_obj_get_type` was inlined and the compile saw it was just a load that distinguished one if branch from another, then I think it will optimise out the load (of the type).

**Comment by andrewleech** [comment]
Ok, did a quick test: `objstr.h`
```
#include "objarray.h"

#define GET_BASE_TYPE(obj_in) \
    ((mp_obj_type_t *)((mp_obj_base_t *)MP_OBJ_TO_PTR(obj_in))->type)

#define GET_STR_DATA_LEN(str_obj_in, str_data, str_len) \
    const byte *str_data; size_t str_len; \
    if (mp_obj_is_qstr(str_obj_in)) \
    { str_data = qstr_data(MP_OBJ_QSTR_VALUE(str_obj_in), &str_len); } \
    else if (GET_BASE_TYPE(str_obj_in) == &mp_type_str || GET_BASE_TYPE(str_obj_in) == &mp_type_bytes) \
    { str_len = ((mp_obj_str_t *)MP_OBJ_TO_PTR(str_obj_in))->len; str_data = ((mp_obj_str_t *)MP_OBJ_TO_PTR(str_obj_in))->data; } \
    else { str_len = ((mp_obj_array_t *)MP_OBJ_TO_PTR(str_obj_in))->len; str_data = ((mp_obj_array_t *)MP_OBJ_TO_PTR(str_obj_in))->items; }

```
on linux, before: 
```
LINK micropython-coverage
   text    data     bss     dec     hex filename
1507301  292304  216040 2015645  1ec19d micropython-coverage
```

after
```
LINK micropython-coverage
   text    data     bss     dec     hex filename
1510349  292304  216648 2019301  1ecfe5 micropython-coverage
```
```

**Comment by dpgeorge** [comment]
I did a similar test with:
```c
#define GET_STR_DATA_LEN(str_obj_in, str_data, str_len) \
    const byte *str_data; \
    size_t str_len; \
    if (mp_obj_is_qstr(str_obj_in)) { \
        str_data = qstr_data(MP_OBJ_QSTR_VALUE(str_obj_in), &str_len); \
    } else if (((mp_obj_base_t *)MP_OBJ_TO_PTR(str_obj_in))->type != &mp_type_bytearray) { \
        str_len = ((mp_obj_str_t *)MP_OBJ_TO_PTR(str_obj_in))->len; \
        str_data = ((mp_obj_str_t *)MP_OBJ_TO_PTR(str_obj_in))->data; \
    } else { \
        str_len = ((mp_obj_array_t *)MP_OBJ_TO_PTR(str_obj_in))->len; \
        str_data = ((mp_obj_array_t *)MP_OBJ_TO_PTR(str_obj_in))->items; \
    }
```
and it adds +16 bytes code on unix x64 and stm32 (with gcc 12.1).  The only reason it increases code size on stm32 is because the code uses r8 instead of r7 in some places (I assume because of the way it optimised out the code), which requires wide Thumb-2 instructions instead of narrow ones (which cost an extra 2 bytes).

**Comment by andrewleech** [comment]
Ah, there's a much bigger discrepancy on unix -coverage than on -dev.  With my old gcc 9.4.0
before:
   text    data     bss     dec     hex filename
 838404   84032   16608  939044   e5424 micropython-dev

yours:
   text    data     bss     dec     hex filename
 838428   84032   16608  939068   e543c micropython-dev

mine:
   text    data     bss     dec     hex filename
 838428   84032   16608  939068   e543c micropython-dev

So yeah that's not anywhere near as expensive as I'd assumed it would be. I'd think it's important to `if type ==`  check for both bytes/str or array/bytearray? Either way mine which is checking both seems to optimise down like you thought.

**Comment by jimmo** [comment]
Can you make a static assert using offsetof?

Or possibly do it with an if statement and when the offsets are the same it'll do the small way, but with fallback to the type check?

**Comment by andrewleech** [comment]
The MP_STATIC_ASSERT was my first thought, however it only works inside a function - not at file scope. So a little annoying. That's why I fell back to the coverage based C test because that'll break the build before letting it get accidentally merged to master at least. Could use the MP_STATIC_ASSERT in there too I guess for even faster feedback.

I still understand it's a bit wrong to use the mp_obj in this macro/function as the wrong struct, that's a dangerous precident to set really.

Alternatively could re-organise both of these type structs to have a common base type that contains the len/data elements. That'd be a much bigger refactor though. Or literally use the same struct for both but `#define` the array element names onto the str struct, similar to the way `free` is renamed for memview.

**Comment by jimmo** [comment]
This is what I meant:

```c
#define GET_STR_DATA_LEN(str_obj_in, str_data, str_len) \
    const byte *str_data; \
    size_t str_len; \
    if (mp_obj_is_qstr(str_obj_in)) { \
        str_data = qstr_data(MP_OBJ_QSTR_VALUE(str_obj_in), &str_len); \
    } else { \
        MP_STATIC_ASSERT(offsetof(mp_obj_str_t, len) == offsetof(mp_obj_array_t, len) && offsetof(mp_obj_str_t, data) == offsetof(mp_obj_array_t, items)); \
        str_len = ((mp_obj_array_t *)MP_OBJ_TO_PTR(str_obj_in))->len; \
        str_data = ((mp_obj_array_t *)MP_OBJ_TO_PTR(str_obj_in))->items; \
    }
```

or with a fallback:

```c
#define GET_STR_DATA_LEN(str_obj_in, str_data, str_len) \
    const byte *str_data; \
    size_t str_len; \
    if (mp_obj_is_qstr(str_obj_in)) { \
        str_data = qstr_data(MP_OBJ_QSTR_VALUE(str_obj_in), &str_len); \
    } else if ((offsetof(mp_obj_str_t, len) == offsetof(mp_obj_array_t, len) && offsetof(mp_obj_str_t, data) == offsetof(mp_obj_array_t, items)) || ((mp_obj_base_t *)MP_OBJ_TO_PTR(str_obj_in))->type != &mp_type_bytearray) { \
        str_len = ((mp_obj_str_t *)MP_OBJ_TO_PTR(str_obj_in))->len; \
        str_data = ((mp_obj_str_t *)MP_OBJ_TO_PTR(str_obj_in))->data; \
    } else { \
        str_len = ((mp_obj_array_t *)MP_OBJ_TO_PTR(str_obj_in))->len; \
        str_data = ((mp_obj_array_t *)MP_OBJ_TO_PTR(str_obj_in))->items; \
    }
```

**Comment by andrewleech** [comment]
Good point @jimmo, if nothing else, adding the offsetof check to the if means the compiler can completely get rid of the second branch with the currently codebase.
``` c 
#include "objarray.h"
#define GET_BASE_TYPE(obj_in) \
    ((mp_obj_type_t *)((mp_obj_base_t *)MP_OBJ_TO_PTR(obj_in))->type)

#define GET_STR_DATA_LEN(str_obj_in, str_data, str_len) \
    const byte *str_data; size_t str_len; \
    if (mp_obj_is_qstr(str_obj_in)) \
    { str_data = qstr_data(MP_OBJ_QSTR_VALUE(str_obj_in), &str_len); } \
    else if ((offsetof(mp_obj_str_t, len) == offsetof(mp_obj_array_t, len) && offsetof(mp_obj_str_t, data) == offsetof(mp_obj_array_t, items)) || \
    (GET_BASE_TYPE(str_obj_in) == &mp_type_str || GET_BASE_TYPE(str_obj_in) == &mp_type_bytes)) \
    { str_len = ((mp_obj_str_t *)MP_OBJ_TO_PTR(str_obj_in))->len; str_data = ((mp_obj_str_t *)MP_OBJ_TO_PTR(str_obj_in))->data; } \
    else { str_len = ((mp_obj_array_t *)MP_OBJ_TO_PTR(str_obj_in))->len; str_data = ((mp_obj_array_t *)MP_OBJ_TO_PTR(str_obj_in))->items; }
```

If the code changed in future though this would still be safe - though it might make it harder to detect if/when that happens.

**Comment by andrewleech** [comment]
actually... I'd possibly prefer this fallback options for one particular use case... https://github.com/micropython/micropython/pull/9028

TLDR on 32bit platforms memoryview cannot have a slice start larger than 16MB. On 32bit (arm) linux use this is actually pretty limiting for me recently. I originally tried to make the `mp_obj_array_t` struct larger in this case so that the `free/start` element was a full `size_t` rather than the current 24bit long... however that seemed to break a lot of my code. I now understand why - this implicit struct match restriction!
So, in the hypothetical case of wanting to fix the memoryview limitation properly (maybe a new `MICROPY_*` setting to increase the size of the  `free/start` element ) it might be better to be able to just fix the `mp_obj_array_t` definition without needing to "artificially" pad out `mp_obj_str_t` as well?

**Comment by dpgeorge** [comment]
Let's just add a simple static assert into the `GET_STR_DATA_LEN` macro.  That will make sure the structs cannot change without a compile error.

I'll make the change during merge.

**Comment by dpgeorge** [comment]
> So, in the hypothetical case of wanting to fix the memoryview limitation properly (maybe a new `MICROPY_*` setting to increase the size of the `free/start` element

We should tackle this separately.  I don't think it's a good idea to increase the size of the struct for memoryview because it'll go beyond the size of a GC block.

**Comment by andrewleech** [comment]
> We should tackle this separately. I don't think it's a good idea to increase the size of the struct for memoryview because it'll go beyond the size of a GC block.

Yes I spoke briefly to @jimmo about this on slack; it's rather rare that 32bit platforms will need to essentially memory map a >16MB file. For my own use case I've got a workaround using uctypes that's also quite performant. As such I think a section about this on the https://docs.micropython.org/en/latest/genrst/modules.html#array page is probably enough, possibly with my workaround snippet as example. I'll endevour to add that as part of https://github.com/micropython/micropython/pull/9028

**Comment by dpgeorge** [comment]
Merged in f7f56d42851aaff2027e23a8ca45c1f1973f1aca with some minor tweaks (to have +0 code size on bare-arm) and clean up.

Code size diff for the final merged version of this is:
```
   bare-arm:    +0 +0.000% 
minimal x86:   +29 +0.018% 
   unix x64:  -792 -0.128% standard[incl -448(data)]
unix nanbox:  -704 -0.127% nanbox[incl -512(data)]
      stm32:   -40 -0.010% PYBV10
     cc3200:   -32 -0.017% 
    esp8266:   -28 -0.004% GENERIC
      esp32:   -72 -0.005% GENERIC[incl -200(data)]
     mimxrt:   -40 -0.011% TEENSY40
 renesas-ra:   -40 -0.006% RA6M2_EK
        nrf:   -16 -0.009% pca10040
        rp2:   -64 -0.013% PICO
       samd:  +148 +0.105% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```
